transformations/transformations/global_var_offload.py

```diff
@@ -0,0 +1,208 @@
+        self._acc_copyout = set()
+        self._var_set = set()
+        self._modules = {}
+
+    def transform_module(self, module, **kwargs):
+        """
+        Add device-side declarations for imported variables.
+        """
+
+        item = kwargs['item']
+
+        # bail if not global variable import
+        if not isinstance(item, GlobalVarImportItem):
+            return
+
+        # confirm that var to be offloaded is declared in module
+        symbol = item.name.split('#')[-1].lower()
+        assert symbol in [s.name.lower() for s in module.variables]
+
+        # do nothing if var is a parameter
+        if module.symbol_map[symbol].type.parameter:
+            return
+
+        # check if var is already declared
+        pragmas = [p for p in FindNodes(Pragma).visit(module.spec) if p.keyword.lower() == 'acc']
+        for p in pragmas:
+            if re.search(fr'\b{symbol}\b', p.content.lower()):
+                return
+
+        # Update the set of variables to be offloaded
+        self._var_set.add(symbol)
+
+        # Write ACC declare pragma
+        module.spec.append(Pragma(keyword='acc', content=f'declare create({symbol})'))
+
+    def transform_subroutine(self, routine, **kwargs):
+        role = kwargs.get('role')
+        if role == 'driver':
+            self.process_driver(routine)
+        elif role == 'kernel':
+            self.process_kernel(routine, **kwargs)
+
+    def process_driver(self, routine):
+        """
+        Add offload and/or copy-back directives for the imported variables.
+        """
+
+        update_device = ()
+        update_host = ()
+
+        # build offload pragmas
+        if self._acc_copyin:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='update device(' + ','.join(self._acc_copyin) + ')'),)
+        if self._enter_data_copyin:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='enter data copyin(' + ','.join(self._enter_data_copyin) + ')'),)
+        if self._enter_data_create:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='enter data create(' + ','.join(self._enter_data_create) + ')'),)
+        if self._exit_data:
+            update_host += as_tuple(Pragma(keyword='acc',
+                                           content='exit data copyout(' + ','.join(self._exit_data) + ')'),)
+        if self._acc_copyout:
+            update_host += as_tuple(Pragma(keyword='acc', content='update self(' + ','.join(self._acc_copyout) + ')'),)
+
+        # replace Loki pragmas with acc data/update pragmas
+        pragma_map = {}
+        for pragma in [pragma for pragma in FindNodes(Pragma).visit(routine.body) if pragma.keyword == 'loki']:
+            if 'update_device' in pragma.content:
+                if update_device:
+                    pragma_map.update({pragma: update_device})
```

Comment:
Why not just `pragma_map[pragma] = update_device`? (Also below...)

Reply:
Agreed, done.